aws/src/main/java/org/apache/iceberg/aws/glue/DynamoLockManager.java

```diff
@@ -0,0 +1,312 @@
+          .exponentialBackoff(acquireIntervalMs(), acquireIntervalMs(), acquireTimeoutMs(), 1)
+          .onlyRetryOn(
+              ConditionalCheckFailedException.class,
+              ProvisionedThroughputExceededException.class,
+              TransactionConflictException.class,
+              RequestLimitExceededException.class,
+              InternalServerErrorException.class)
+          .run(id -> acquireOnce(id, ownerId));
+      return true;
+    } catch (RuntimeException e) {
+      return false;
+    }
+  }
+
+  @VisibleForTesting
+  void acquireOnce(String entityId, String ownerId) {
+    GetItemResponse response = dynamo.getItem(GetItemRequest.builder()
+        .tableName(lockTableName)
+        .key(toKey(entityId))
+        .build());
+
+    if (!response.hasItem()) {
+      dynamo.putItem(PutItemRequest.builder()
+          .tableName(lockTableName)
+          .item(toNewItem(entityId, ownerId))
+          .conditionExpression(CONDITION_LOCK_ENTITY_NOT_EXIST)
+          .build());
+    } else {
+      Map<String, AttributeValue> currentItem = response.item();
+
+      try {
+        Thread.sleep(Long.parseLong(currentItem.get(COL_LEASE_DURATION_MS).n()));
+      } catch (InterruptedException e) {
+        throw new IllegalStateException(
+            String.format("Fail to acquire lock %s by %s, interrupted during sleep", entityId, ownerId), e);
+      }
+
+      Map<String, AttributeValue> expressionValues = Maps.newHashMap();
+      expressionValues.put(":eid", AttributeValue.builder().s(entityId).build());
+      expressionValues.put(":vid", AttributeValue.builder().s(currentItem.get(COL_VERSION).s()).build());
+      dynamo.putItem(PutItemRequest.builder()
+          .tableName(lockTableName)
+          .item(toNewItem(entityId, ownerId))
+          .conditionExpression(CONDITION_LOCK_ENTITY_NOT_EXIST_OR_VERSION_MATCH)
+          .expressionAttributeValues(expressionValues)
+          .build());
+    }
+
+    startNewHeartbeat(entityId, ownerId);
+  }
+
+  private void startNewHeartbeat(String entityId, String ownerId) {
+    if (heartbeats.containsKey(entityId)) {
+      heartbeats.remove(entityId).cancel(false);
+    }
+
+    heartbeats.put(entityId, scheduler().scheduleAtFixedRate(() -> dynamo.putItem(PutItemRequest.builder()
+        .tableName(lockTableName)
+        .item(toNewItem(entityId, ownerId))
+        .conditionExpression(CONDITION_LOCK_ID_MATCH)
+        .expressionAttributeValues(toLockIdValues(entityId, ownerId))
+        .build()), 0, heartbeatIntervalMs(), TimeUnit.MILLISECONDS));
+  }
+
+  @Override
+  public boolean release(String entityId, String ownerId) {
+    ScheduledFuture<?> heartbeat = heartbeats.get(entityId);
+    try {
+      dynamo.deleteItem(DeleteItemRequest.builder()
+          .tableName(lockTableName)
+          .key(toKey(entityId))
+          .conditionExpression(CONDITION_LOCK_ID_MATCH)
+          .expressionAttributeValues(toLockIdValues(entityId, ownerId))
+          .build());
```

Comment:
I think this method is a good candidate to use `ExceptionUtil.runSafely`.

Cancelling the heartbeat should be done in a `finally` block in case there is a connection issue that causes this to throw an exception. Otherwise, the heartbeat thread could accidentally hold the lock forever if it succeeds.

The reason I would say to use `runSafely` is that I think it would be better to automatically suppress any exceptions in the `finally` block and that's what `runSafely` does. I doubt that `heartbeat.cancel(false)` would throw any exceptions, but it's good to be careful in case the cancellation code changes in the future.

Reply:
I think we cannot always cancel heartbeat in a final block. If the caller has a different owner, the heartbeat should not be canceled. Therefore, the correct behavior should be to check the owner of the lock when trying to release, so that:
1. if release succeeds, then cancel heartbeat
2. if release fails and heartbeat owner is the same, cancel heartbeat. There should be something wrong with Dynamo, and canceling the heartbeat should allow the lock to expire
3. if release fails and heartbeat owner is different, the caller is wrong and we should not cancel the ongoing heartbeat.

To achieve this, I have introduced an internal class `DynamoHeartbeat` to also save the owner in the `heartbeats` map.

Also, I added retry in release using `Tasks` to retry on some retryable exceptions.

Reply:
This looks good. Nice catch with the owner check.